### DISCUSSION TITLE
Converting 3D files (PLY / STL / glTF) to IFC ?

### DISCUSSION THREAD

**Discussion by swiss-knight** [IfcOpenShell/IfcOpenShell#4503] [Q&A]
Hello,

I have some geometric 3D GIS data that I can export to common 3D file formats such as PLY, STL, GLB/glTF.

99% of my researches so far get me to conversions stories from IFC -> glTF but never the opposite.

Hence my questions :
do you know where I can start to encapsulate / embed / convert those geometries into some valid IFC files (not sure which verb I should use) ?
What would be the best 3D format to be able to gently step in the IFC world ?

Thanks.

**Comment by Moult** [comment]
See https://blenderbim.org/docs-python/introduction/introduction_to_bim.html

> BIM data is very different from a regular 3D model. In fact, geometry is optional, and most data is non-geometric. This means that it is not simply a 3D format that you can import or export from and expect meaningful results.

**Reply by swiss-knight** [reply]
@Moult
Do you know where the https://blenderbim.org/docs-python/introduction/introduction_to_bim.html page has been moved to?
Because it is currently:
1) redirected to: https://bonsaibim.org/docs-python/introduction/introduction_to_bim.html
2) raising a `404` error.

🙏Thanks!

**Reply by Moult** [reply]
https://docs.bonsaibim.org/quickstart/introduction_to_bim.html

**Comment by civilx64** [comment]
Dion is correct; BIM data is very different than pure geometry models like PLY / STL / glTF that you mentioned.  However given the fact that these geometries are coming from GIS then there is a reasonable chance of formatting that into a reasonable equivalent in IFC.

In my experience, IFC model data typically fall into three main categories: products, representations, and metadata.  Products can be generally thought of as physical objects such as doors, windows, and walls in a building or roads, railways, and bridges in an outdoor context.  Representations refer to the geometry of a product.  Metadata consists of things like names, descriptions, property sets, quantity takeoffs, and material (e.g. wood, concrete, steel) information.

On the GIS side I am accustomed to terminology where a feature class roughly equates to a table in a relational database, and each feature consists of geometry + attributes.  This leads to an approximate mapping of GIS -> IFC as follows:

| GIS | IFC |
|---|------------|
|Feature Class | IfcProduct |
|Geometry | IfcShapeRepresentation |
|Attributes | IfcPropertySet|

The given that the 3D formats mentioned all involve triangulated meshes then I assume that is the geometry type of the GIS features that you are looking at - which would map to [IfcTriangulatedFaceSet](https://standards.buildingsmart.org/IFC/RELEASE/IFC4_1/FINAL/HTML/schema/ifcgeometricmodelresource/lexical/ifctriangulatedfaceset.htm).  Depending on your feature classes you may be able to map them to a semantically-meaningful IfcProduct but my hunch is that you would need to fall back to something generic like [IfcBuildingElementProxy](https://standards.buildingsmart.org/IFC/RELEASE/IFC4_1/FINAL/HTML/schema/ifcsharedbldgelements/lexical/ifcbuildingelementproxy.htm). 
As an example, I commonly see GIS feature classes for traffic signals that could be mapped to the [IfcSignal](https://standards.buildingsmart.org/IFC/RELEASE/IFC4_3/HTML/lexical/IfcSignal.htm) product,  

 Your GIS attributes will take the form of a custom [IfcPropertySet](https://standards.buildingsmart.org/IFC/RELEASE/IFC4_1/FINAL/HTML/schema/ifckernel/lexical/ifcpropertyset.htm) where you have the ability to enforce the data types of your GIS data dictionary and can map your domains to enumerated property sets to meet data governance requirements.

To answer your question - I don't think there is much difference between targeting PLY versus STL versus glTF for the geometry.  One potential benefit of glTF is that you can attach metadata to the meshes in the scene but my hunch is you would be better using glTF for geometry only and lean on something else like [FME from Safe Software](https://www.safe.com/) to ETL the GIS to IFC.  Or you could hire a software development consultancy to assist, depending on how "commercial" your use case is and how much data you are looking to potentially convert.

Hope that helps.  Good luck!

**Comment by Moult** [comment]
The other two issues you'll face to get a minimally useful IFC is 1) to create a spatial tree and 2) to decide which IFC class to use for the elements. As a worst case scenario you can put everything in a single building / site and use generic proxies for everything but it's not ideal.

**Comment by luxiousCoder** [comment]
https://github.com/andyward/Obj2Ifc  
Try it!
